Add OomScoreAdj

runcom opened the pull request:
libcontainer v0.0.4 introduces setting `/proc/self/oom_score_adj` to
better tune oom killing preferences for container process. This patch
simply integrates OomScoreAdj libcontainer's config option and adjust
the cli with this new option.

close #13116

TODOs:
- [x] Docs
- [x] Tests

Signed-off-by: Antonio Murdaca runcom@linux.com

RichardScothern commented on runconfig/parse.go:
```diff
@@ -71,6 +71,7 @@ func Parse(cmd *flag.FlagSet, args []string) (*Config, *HostConfig, *flag.FlagSe
 		flStdin           = cmd.Bool([]string{"i", "-interactive"}, false, "Keep STDIN open even if not attached")
 		flTty             = cmd.Bool([]string{"t", "-tty"}, false, "Allocate a pseudo-TTY")
 		flOomKillDisable  = cmd.Bool([]string{"-oom-kill-disable"}, false, "Disable OOM Killer")
+		flOomScoreAdj     = cmd.Int([]string{"-oom-score-adj"}, 0, "Tune container OOM score adjustment")
```
Should this description contain the accepted range (-1000, 1000)?

RichardScothern commented on runconfig/parse.go:
```diff
@@ -71,6 +71,7 @@ func Parse(cmd *flag.FlagSet, args []string) (*Config, *HostConfig, *flag.FlagSe
 		flStdin           = cmd.Bool([]string{"i", "-interactive"}, false, "Keep STDIN open even if not attached")
 		flTty             = cmd.Bool([]string{"t", "-tty"}, false, "Allocate a pseudo-TTY")
 		flOomKillDisable  = cmd.Bool([]string{"-oom-kill-disable"}, false, "Disable OOM Killer")
```
Is this still necessary with this change?

runcom replied:
oom-kill-disable is per cgroup config, oom-score-adj is per process. Also oom-score-adj will weight oom killer preferences when looking for procs to kill

icecrime commented:
Design OK

icecrime commented:
Ping @runcom for rebase before code review!

runcom commented:
@icecrime rebased

vdemeester commented:
@runcom looks like the `vet` is not happy :sweat_smile:

runcom committed:
Add OomScoreAdj to configure container oom killer preferences

libcontainer v0.0.4 introduces setting `/proc/self/oom_score_adj` to
better tune oom killing preferences for container process. This patch
simply integrates OomScoreAdj libcontainer's config option and adjust
the cli with this new option.

Signed-off-by: Antonio Murdaca <amurdaca@redhat.com>
Signed-off-by: Antonio Murdaca <runcom@redhat.com>

runcom commented:
rebase hell :/ updated

thaJeztah commented:
needs another rebase :relaxed:

runcom commented:
rebased again

icecrime commented:
LGTm

runcom commented:
rebased (again)

thaJeztah commented:
ping @crosbymichael @phemmer (think you may be interested having a look at this as well :+1:)

phemmer commented:
Seems a pretty simple implementation to me.
Thanks :-)

runcom commented:
rebased!

thaJeztah commented:
ping @runcom could you rebase once more, then I'll try to kick @crosbymichael to give this another review :smile:

runcom commented:
6th rebase done :smile:

crosbymichael commented:
is this expected? FAIL: docker_cli_events_unix_test.go:48: DockerSuite.TestEventsOOMDisableFalse

vishh commented:
Thanks for this PR @runcom!

runcom commented:
@crosbymichael pretty sure it doesn't

thaJeztah commented:
It's all green now!

LK4D4 commented:
@runcom need rebase

runcom commented:
@LK4D4 rebased

LK4D4 commented:
LGTM

LK4D4 commented:
ping @thaJeztah @moxiegirl

moxiegirl commented on docs/reference/api/docker_remote_api.md:
```diff
@@ -104,11 +104,11 @@ This section lists each version from latest to oldest.  Each listing includes a
 * `DELETE /volumes/(name)`remove a volume with the specified name.
-* `VolumeDriver` has been moved from config to hostConfig to make the configuration portable.
-* `GET /images/(name)/json` now returns information about tags and digests of the image.
+* `VolumeDriver` has been moved from config to HostConfig to make the configuration portable.
```
```markdown
* `VolumeDriver` was moved from `config` to `HostConfig` to make the configuration portable.
```

moxiegirl commented on docs/reference/api/docker_remote_api.md:
```diff
@@ -104,11 +104,11 @@ This section lists each version from latest to oldest.  Each listing includes a
 * The `config` option now accepts the field `StopSignal`, which specifies the signal to use to kill a container.
 * `GET /containers/(id)/stats` will return networking information respectively for each interface.
-* The `hostConfig` option now accepts the field `DnsOptions`, which specifies a
+* The `HostConfig` option now accepts the field `DnsOptions`, which specifies a
```
```markdown
* The `HostConfig` option now includes the `DnsOptions` field to configure a
the container's DNS options.
```

moxiegirl commented on docs/reference/api/docker_remote_api.md:
```diff
@@ -127,6 +127,7 @@ list of DNS options to be used in the container.
   `NetworkSettings.Gateway`, `NetworkSettings.IPAddress`,
   `NetworkSettings.IPPrefixLen`, and `NetworkSettings.MacAddress` fields, which
   are still returned for backward-compatibility, but will be removed in a future version.
+* The `HostConfig` option now accepts the field `OomScoreAdj`, which is used to adjust the badness heuristic used to select which process gets killed in out-of-memory conditions by the OOM killer.
```
```markdown
* The `HostConfig` option now includes the `OomScoreAdj` field for adjusting the
badness heuristic. This heuristic selects which processes the OOM killer kills
under out-of-memory conditions.
```

moxiegirl commented on docs/reference/commandline/create.md:
```diff
@@ -57,6 +57,7 @@ Creates a new container.
       --name=""                     Assign a name to the container
       --net="default"               Set the Network mode for the container
       --oom-kill-disable=false      Whether to disable OOM Killer for the container or not
+      --oom-score-adj=0             Tune OOM preferences for container (-1000 to 1000)
```
```markdown
Tunes container's OOM preferences (accepts -1000 to 1000)
```

moxiegirl commented on docs/reference/commandline/run.md:
```diff
@@ -61,6 +61,7 @@ parent = "smn_cli"
                                     'host': use the host network stack inside the container
                                     'NETWORK': connects the container to user-created network using `docker network create` command
       --oom-kill-disable=false      Whether to disable OOM Killer for the container or not
+      --oom-score-adj=0             Tune OOM preferences for container (-1000 to 1000)
```
```markdown
Tunes container's OOM preferences (accepts -1000 to 1000)
```

runcom replied:
does this imply the OOM is owned/inside the container? where instead it's in the kernel (host)

moxiegirl replied:
Tune's the host's OOM preferences for containers. (accepts -1000 to 1000)

runcom replied:
thanks!

thaJeztah commented:
ping @runcom can you rebase and address @moxiegirl's comments?

runcom commented:
@thaJeztah sorry haven't had time, will rebase and update the PR tomorrow morning!

runcom commented:
@moxiegirl @thaJeztah PTAL

thaJeztah commented on docs/reference/api/docker_remote_api.md:
```diff
@@ -107,12 +107,11 @@ This section lists each version from latest to oldest.  Each listing includes a
-* `VolumeDriver` has been moved from config to hostConfig to make the configuration portable.
-* `GET /images/(name)/json` now returns information about tags and digests of the image.
+* `VolumeDriver` was moved from `config` to `HostConfig` to make the configuration portable.
+* `GET /images/(name)/json` now returns information about image tags.
```
Oh, I see this was suggested by @moxiegirl, but this line deliberately mentioned "tags" and "digests", because both are included (`RepoTags`, `RepoDigests`)

@moxiegirl any reason for dropping "digests"?

runcom replied:
@moxiegirl @thaJeztah can you also re-evaluate https://github.com/docker/docker/pull/16277#discussion_r45956486? :)

moxiegirl replied:
`GET /images/(name)/json` now returns information about an image's `RepoTags` and `RepoDigests`.

thaJeztah replied:
works for me, thanks!

thaJeztah commented on runconfig/parse.go:
```diff
@@ -79,6 +79,7 @@ func Parse(cmd *flag.FlagSet, args []string) (*Config, *HostConfig, *flag.FlagSe
 		flStdin             = cmd.Bool([]string{"i", "-interactive"}, false, "Keep STDIN open even if not attached")
 		flTty               = cmd.Bool([]string{"t", "-tty"}, false, "Allocate a pseudo-TTY")
 		flOomKillDisable    = cmd.Bool([]string{"-oom-kill-disable"}, false, "Disable OOM Killer")
+		flOomScoreAdj       = cmd.Int([]string{"-oom-score-adj"}, 0, "Tune OOM preferences for container (-1000 to 1000)")
```
Can we use the same text here? `Tunes OOM preferences for the container (accepts -1000 to 1000)` or will the length be > 80 chars then?

thaJeztah commented on contrib/completion/zsh/_docker:
```diff
@@ -482,6 +482,7 @@ __docker_subcommand() {
         "($help)--name=[Container name]:name: "
         "($help)--net=[Connect a container to a network]:network mode:(bridge none container host)"
         "($help)--oom-kill-disable[Disable OOM Killer]"
+        "($help)--oom-score-adj[Tune OOM preferences for container (-1000 to 1000)]"
```
Can you use the same text here as well?

thaJeztah commented:
@moxiegirl I see we're inconsistent with describing the options; `Tunes ... something` v.s. `Tune ... something`. I think we should pick one direction, so either *"(this option) does something"*, or *"(use this option to) do something"*; wdyt?

moxiegirl commented:
@thaJeztah Yes, I noticed that. I think that is a good idea but the problem is beyond the scope of this PR.  It is really something we should do in another PR. WDYT?

thaJeztah commented:
@moxiegirl I think so far we used `Tune...` but your suggestions here were `Tunes..`, that was my reason for mentioning it. But if there are other inconsistencies, I'm perfectly fine changing it in a separate PR

moxiegirl commented:
Ah, I see. Yes, you are right `Tune` is more consistent with the other options.

runcom commented:
@moxiegirl @thaJeztah updated (I'll leave what's not consistent to a follow up PR as you agreed on)

thaJeztah commented:
@runcom looks like I overlooked that changes to the man-pages are missing, can you add the flag to the man-pages for `create` and `run`? :blush:

runcom commented:
@thaJeztah added

thaJeztah commented:
Thanks! And thanks for being patient with us :+1: 

LGTM

ping @moxiegirl for a final look

runcom commented:
ping @moxiegirl O:)

moxiegirl commented:
@runcom LGTM to you! :-D Thank you. @thaJeztah for the merge
